Extension/lib/settings/sync-service.js

```diff
@@ -0,0 +1,265 @@
+     * corresponding to the direction: remote-to-local or local-to-remote.
+     *
+     * After that we load selected data sections and process the compatibility result by processSections method,
+     * simply saving the newer sections to remote or local storage.
+     *
+     * As the end we updated manifests with processSections result, setting up the sections timestamps.
+     *
+     * @param callback
+     */
+    var processManifest = function (callback) {
+        var onManifestLoaded = function (remoteManifest) {
+            console.log('Processing remote manifest..');
+
+            if (!remoteManifest) {
+                callback();
+                return;
+            }
+
+            var localManifest = SettingsProvider.loadSettingsManifest();
+
+            var compatibility = findCompatibility(remoteManifest, localManifest);
+            console.log(compatibility);
+            if (!compatibility.canRead) {
+                console.log('Protocol versions are not compatible');
+                callback();
+                return;
+            }
+
+            processSections(localManifest, remoteManifest, compatibility, function (updatedSections) {
+                console.log(updatedSections);
+
+                var updatedDate = new Date().getTime();
+                if (updatedSections.remoteToLocal.length > 0) {
+                    console.log('Saving local manifest..');
+
+                    updateManifestSections(localManifest.sections, updatedSections.remoteToLocal);
+
+                    localManifest.timestamp = updatedDate;
+                    SettingsProvider.saveSettingsManifest(localManifest);
+                }
+
+                if (compatibility.canWrite) {
+                    console.log('Saving remote manifest..');
+
+                    updateManifestSections(remoteManifest.sections, updatedSections.localToRemote);
+
+                    remoteManifest.timestamp = updatedDate;
+                    remoteManifest["app-id"] = localManifest["app-id"];
+
+                    SyncProvider.save(MANIFEST_PATH, remoteManifest, callback);
+                } else {
+                    callback();
+                }
+            });
+        };
+
+        console.log('Loading remote manifest..');
+        SyncProvider.load(MANIFEST_PATH, onManifestLoaded);
+    };
+
+    /**
+     * Processes sections synchronization.
+     * Return object with updates sections info.
+     *
+     * @param localManifest
+     * @param remoteManifest
+     * @param compatibility
+     * @param callback
+     */
+    var processSections = function (localManifest, remoteManifest, compatibility, callback) {
+        var updatedSections = {
+            localToRemote: [],
+            remoteToLocal: []
+        };
+
+        var dfds = [];
+        var dfd;
+
+        for (var i = 0; i < compatibility.sectionsLocalToRemote.length; i++) {
```

Comment:
We can get rid of the callbacks hell here if you make our interface functions returning `Promise` object instead of taking `callback` as a parameter.

This will also help you with error handling.